sockeye/beam_search.py

```diff
@@ -795,15 +807,19 @@ def forward(self,
             scores, lengths = self._update_scores(target_dists, finished, scores_accumulated,
                                                   lengths, max_output_lengths, pad_dist, eos_dist)
 
+            if target_prefix is not None and t <= target_prefix.size(-1):
+                target_prefix_in_one_hot = utils.one_hot_encoding_from_target_prefix(target_prefix[:,t-1:t], output_vocab_size)
+                target_prefix_in_one_hot = target_prefix_in_one_hot.expand(-1, self.beam_size, -1).reshape(-1, target_prefix_in_one_hot.size(-1))
+                scores.masked_fill_(target_prefix_in_one_hot == 0, onp.inf)
             # (3) Get beam_size winning hypotheses for each sentence block separately. Only look as
             # far as the active beam size for each sentence.
             if self._sample is not None:
                 best_hyp_indices, best_word_indices, scores_accumulated = self._sample(scores, target_dists, finished)
             else:
                 # On the first timestep, all hypotheses have identical histories, so force topk() to choose extensions
                 # of the first row only by setting all other rows to inf
-                if t == 1:
-                    scores += first_step_mask
+                if t <= first_step_mask.size(-1):
```

Comment:
This is critical path code for performance and this change will request the size of a tensor repeatedly, as well as require additional indexing on `first_step_mask` below.
Can we change the logic to keep the existing code for `target_prefix == None`?

It is a bit unfortunate that we are adding more data-dependent control flow here, it doesn't easily allow us to fold this into a traced module.